src/machine/asm/execute.S

```diff
@@ -96,11 +138,62 @@
 #define MOV_RS2r_TO_RDX movq RS2r, %rdx
 #define MOV_RDX_TO_RS2r movq %rdx, RS2r
 
-#define CHECK_WRITE_PERMISSION(address_reg, temp_reg1, temp_reg2, temp_reg2d, length) \
+#ifdef __APPLE__
+#define CALL_MEMSET call _memset
+#elif defined IS_WINDOWS
+#define CALL_MEMSET call memset
+#else
+#define CALL_MEMSET call memset@plt
+#endif
+
+
+/*
+ * Same as access_control in src/machine/aot/aot.x64.c
+ */
+#define ACCESS_CONTROL(address_reg, temp_reg1, temp_reg2, temp_reg2d, length, check_write) \
   movq address_reg, temp_reg1; \
-  shr $CKB_VM_ASM_RISCV_PAGE_SHIFTS, temp_reg1; \
-  cmp $CKB_VM_ASM_RISCV_PAGES, temp_reg1; \
+  shr $CKB_VM_ASM_MEMORY_FRAME_SHIFTS, temp_reg1; \
+  cmp $CKB_VM_ASM_MEMORY_FRAMES, temp_reg1; \
+  jae .exit_out_of_bound; \
+  movzbl CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_FRAMES(MACHINE, temp_reg1), temp_reg2d; \
+  cmp $1, temp_reg2d; \
+  je 1f; \
+  movb $1, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_FRAMES(MACHINE, temp_reg1); \
+  PREPCALL; \
+  movq temp_reg1, ARG1; \
+  shl $CKB_VM_ASM_MEMORY_FRAME_SHIFTS, ARG1; \
+  addq $CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY, ARG1; \
+  addq MACHINE, ARG1; \
+  xor ARG2, ARG2; \
+  movq $CKB_VM_ASM_MEMORY_FRAMESIZE, ARG3; \
+  CALL_MEMSET; \
+  POSTCALL; \
+1: \
+  movq address_reg, temp_reg1; \
+  addq $length, temp_reg1; \
+  subq $1, temp_reg1; \
+  shr $CKB_VM_ASM_MEMORY_FRAME_SHIFTS, temp_reg1; \
+  cmp $CKB_VM_ASM_MEMORY_FRAMES, temp_reg1; \
   jae .exit_out_of_bound; \
+  movzbl CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_FRAMES(MACHINE, temp_reg1), temp_reg2d; \
+  cmp $1, temp_reg2d; \
+  je 2f; \
+  movb $1, CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_FRAMES(MACHINE, temp_reg1);\
+  PREPCALL; \
+  movq temp_reg1, ARG1; \
+  shl $CKB_VM_ASM_MEMORY_FRAME_SHIFTS, ARG1; \
+  addq $CKB_VM_ASM_ASM_CORE_MACHINE_OFFSET_MEMORY, ARG1; \
+  addq MACHINE, ARG1; \
+  xor ARG2, ARG2; \
+  movq $CKB_VM_ASM_MEMORY_FRAMESIZE, ARG3; \
+  CALL_MEMSET; \
+  POSTCALL; \
+2: \
+  movq $check_write, temp_reg1; \
```

Comment:
I feel it is better if we define 2 macros here, one deals with both memory initialization and write permission check, for store operations. One deals only with memory initialization for load operations.

The problem here, is that even though we do a comparison based on `check_write`, and skip the operations in the read calls, those will still be generated in the code, resulting in larger code size. For a hot loop that is extremely performance sensitive, I suggest we keep the code as short as possible, even though we might experience some code duplication

Reply:
I rolled back the modification of this file